funsor/distributions2.py

```diff
@@ -0,0 +1,194 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+from collections import OrderedDict
+
+import makefun
+import torch
+import pyro.distributions as dist
+from pyro.distributions.torch_distribution import MaskedDistribution
+from pyro.distributions.util import broadcast_shape
+
+import funsor.ops as ops
+from funsor.domains import Domain, bint, reals
+from funsor.tensor import Tensor, align_tensors
+from funsor.terms import Funsor, FunsorMeta, Independent, Number, Variable, eager, to_data, to_funsor
+
+
+def _dummy_tensor(domain):
+    return torch.tensor(0.1 if domain.dtype == 'real' else 1).expand(domain.shape)
+
+
+class DistributionMeta2(FunsorMeta):
+    def __call__(cls, *args, name=None):
+        if len(args) < len(cls._ast_fields):
+            args = args + (name if name is not None else 'value',)
+        return super(DistributionMeta2, cls).__call__(*args)
+
+
+class Distribution2(Funsor, metaclass=DistributionMeta2):
+    """
+    Different design for the Distribution Funsor wrapper,
+    closer to Gaussian or Delta in which the value is a fresh input.
+    """
+    dist_class = dist.Distribution  # defined by derived classes
+
+    def __init__(self, *args, name='value'):
+        params = OrderedDict(zip(self._ast_fields, args))
+        inputs = OrderedDict()
+        for param_name, value in params.items():
+            assert isinstance(param_name, str)
+            assert isinstance(value, Funsor)
+            inputs.update(value.inputs)
+        assert isinstance(name, str) and name not in inputs
+        inputs[name] = self._infer_value_shape(**params)
+        output = reals()
+        fresh = frozenset({name})
+        bound = frozenset()
+        super().__init__(inputs, output, fresh, bound)
+        self.params = params
+        self.name = name
+
+    def __getattribute__(self, attr):
+        if attr in type(self)._ast_fields and attr != 'name':
+            return self.params[attr]
+        return super().__getattribute__(attr)
+
+    @classmethod
+    def _infer_value_shape(cls, **kwargs):
+        # rely on the underlying distribution's logic to infer the event_shape
+        instance = cls.dist_class(**{k: _dummy_tensor(v.output) for k, v in kwargs.items()})
+        out_shape = instance.event_shape
+        if isinstance(instance.support, torch.distributions.constraints._IntegerInterval):
+            out_dtype = instance.support.upper_bound + 1
+        else:
+            out_dtype = 'real'
+        return Domain(dtype=out_dtype, shape=out_shape)
```

Comment:
This method is the main innovation in this PR: it turns out we can reuse the distributions' existing `event_shape` and `support` inference logic to generically infer the shape of the fresh `value` input, rather than reimplementing that logic by hand for every distribution we add.